tools/clone-tutorial.py

```diff
@@ -0,0 +1,268 @@
+    fm['stack'] = new_type
+    fm['static_data'] = new_name
+    permalink = fm['permalink']
+    if new_permalink is None or (new_permalink is not None and (orig_name == new_name)):
+        fm['permalink'] = permalink.replace(old_type, new_type)
+    elif new_permalink is not None and (orig_name != new_name):
+        fm['permalink'] = '/' + new_permalink + '/' + new_type
+
+    new_path = "../tutorials/{name}".format(name=new_name)
+
+    if not os.path.exists(new_path):
+        os.mkdir(new_path)
+
+    with open(new_path + "/{tutorial_type}.html".format(tutorial_type=new_type), 'w') as sf:
+        sf.write('---\n')
+        for key, value in fm.items():
+            sf.write(key + ': ' + value + '\n')
+        sf.write('---\n')
+
+
+def handle_harness_files(orig_single_type,
+                         new_single_type,
+                         test_hrns_dir,
+                         orig_tut_name,
+                         new_tut_name):
+    orig_harness_file = test_hrns_dir + '/' + orig_tut_name + '/' + orig_single_type + '.yml'
+    new_harness_file = test_hrns_dir + '/' + new_tut_name + '/' + new_single_type + '.yml'
+    if not os.path.exists(test_hrns_dir + '/' + new_tut_name):
+        os.mkdir(test_hrns_dir + '/' + new_tut_name)
+    shutil.copy(orig_harness_file, new_harness_file)
+    replace_names_in_file(new_harness_file, args.orig_name, args.new_name)
+    if args.orig_main_class is not None and args.new_main_class is not None:
+        replace_names_in_file(new_harness_file, args.orig_main_class, args.new_main_class)
+        replace_names_in_file(new_harness_file, args.orig_main_class + 'Test', args.new_main_class + 'Test')
+    return [new_single_type]
+
+
+def ensure_valid_tutorial_names():
+    if os.path.exists(proposed_tutorial):
+        print("A tutorial for %s exists" % proposed_tutorial)
+        exit(1)
+    if orig_tutorial_type == '' or orig_tutorial_type not in tutorial_types:
+        print("Tutorial type %s is either unknown or must be provided" % orig_tutorial_type)
+        exit(1)
+    elif new_tutorial_type == '' or new_tutorial_type not in tutorial_types:
+        print("Tutorial type %s is either unknown or must be provided" % new_tutorial_type)
+        exit(1)
+
+
+semaphore_path = '../.semaphore/semaphore.yml'
+tutorials_yaml_path = '../_data/tutorials.yaml'
+tutorials_dir = kt_home + '/_includes/tutorials'
+test_harness_dir = kt_home + '/_data/harnesses'
+orig_name_parts = str(args.orig_name).split('/')
+orig_tutorial_name = orig_name_parts[0]
+new_tutorial_name_parts = str(args.new_name).split('/')
+new_tutorial_name = new_tutorial_name_parts[0]
+orig_tutorial_type = ''
+new_tutorial_type = ''
+if len(orig_name_parts) > 1:
+    orig_tutorial_type = orig_name_parts[1]
+if len(new_tutorial_name_parts) > 1:
+    new_tutorial_type = new_tutorial_name_parts[1]
+if args.debug:
+    print("Names of tutorial to be cloned " + str(orig_name_parts) + '  ' + orig_tutorial_name
```

Comment:
style preference / I think more pythonic to interpolate with f-string:
```
print(f"Names of tutorial to be cloned {orig_name_parts} {orig_tutorial_name} {orig_tutorial_type}")
```
there are a few other spots where I think this'd look better than `+`'ing strings or using `.format()`